skipList/skipList/SkipList.cs

```diff
@@ -0,0 +1,409 @@
+﻿// <copyright file="SkipList.cs" company="Kalinin Andrew">
+// Copyright (c) Kalinin Andrew. All rights reserved.
+// </copyright>
+
+namespace SkipList;
+
+using System.Collections;
+using System.Runtime.ExceptionServices;
+
+/// <summary>
+/// An implementation of a skip list that supports the IListT interface.
+/// </summary>
+/// <typeparam name="T">The type of elements in the collection.</typeparam>
+public class SkipList<T> : IList<T>
+{
+    private const int MaxLevel = 32;
+    private const double Probability = 0.5;
+
+    private readonly Random random = new Random();
+    private readonly IComparer<T> valueComparer;
+    private readonly Node head = new Node(default, MaxLevel);
+
+    private int currentLevels = 1;
+    private int count;
+    private int version;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="SkipList{T}"/> class.
+    /// </summary>
+    public SkipList()
+        : this(Comparer<T>.Default)
+    {
+    }
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="SkipList{T}"/> class.
+    /// </summary>
+    /// <param name="comparer">A comparator for comparing elements.</param>
+    public SkipList(IComparer<T> comparer)
+    {
+        if (comparer == null)
+        {
+            throw new ArgumentNullException(nameof(comparer));
+        }
+
+        this.valueComparer = comparer;
+    }
+
+    /// <summary>
+    /// Gets number of items in the list.
+    /// </summary>
+    public int Count => this.count;
+
+    /// <summary>
+    /// Gets a value indicating whether the list is read-only.
+    /// </summary>
+    public bool IsReadOnly => false;
+
+    /// <summary>
+    /// Gets or sets the element by the specified index.
+    /// </summary>
+    /// <param name="index">The element's index.</param>
+    /// <returns>The element by index.</returns>
+    /// <exception cref="ArgumentOutOfRangeException">It is thrown if the index is out of range.</exception>
+    /// <exception cref="NotImplementedException">It is thrown when trying to set a value.</exception>
+    public T this[int index]
+    {
+        get
+        {
+            if (index < 0 || index >= this.count)
+            {
+                throw new ArgumentOutOfRangeException();
+            }
+
+            return this.GetNodeAtIndex(index).Value;
+        }
+        set => throw new NotImplementedException();
+    }
+
+    /// <summary>
+    /// Adds an item to the list.
+    /// </summary>
+    /// <param name="item">The element to add.</param>
+    public void Add(T item)
+    {
+        Node[] update = new Node[MaxLevel];
+        int[] index = new int[MaxLevel];
```

Comment:
```suggestion
        var update = new Node[MaxLevel];
        var index = new int[MaxLevel];
```
И ниже. Не пишите один тип дважды